Dphan/add mqtt tests

dorotaphanSiili opened the pull request:
<!-- 
Thanks for creating this pull request 🤗

Please make sure that the pull request is limited to one type (docs, feature, etc.) and keep it as small as possible. You can open multiple prs instead of opening a huge one.
-->

<!-- If this pull request closes an issue, please mention the issue number below -->
Closes # <!-- Issue # here -->

## 📑 Description
<!-- Add a brief description of the pr -->

<!-- You can also choose to add a list of changes and if they have been completed or not by using the markdown to-do list syntax
- [ ] Not Completed
- [x] Completed
-->

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

## ℹ Additional Information
<!-- Any additional information like breaking changes, dependencies added, screenshots, comparisons between new and old behavior, etc. -->

dorotaphanSiili committed:
feat(mqtt): add generated tests

For each test interface we generate the mqtt tests.
Tests setup a client and a server and interface adapters.
They do not provide a broker which has to be started before starting
tests.
The test are using non-default generated value for singals and
properties.
As results of methods cannot be predicted their result is not checked.
They are just executed, and waited for result to be delivered.

This is commit fo #3 task

dorotaphanSiili committed:
feat(mqtt): increase safety for tear down client

C Mqtt client is now stored in unique_ptr with custom deleter.
Now when cwrapper is deleted at any point the CMqtt client will be
properly released.
There is also added a check if after 1000ms the request to do anything
is still valid (checking a weak_ptr)
This task is part of #3 task, which require often setup and teardown for
same client.

dorotaphanSiili committed:
fix(mqtt): fix client side topics to match service

This is part of #3 task which provides mqtt tests for generated
interfaces.

dorotaphanSiili committed:
fix(mqtt): void functions do not wait for response

the server side didn't respond for void messages, but the client was
waiting for it. Now the client, for void functions just sends the
request and doesn't wait for the result.

This is part of #3 task.

dorotaphanSiili committed:
feat(mqtt): track subscription state

This commit introduces a base cass for interface adapters,
which purpose is to take care of subscriptions, its state and allow users of
the interface subscribe for events:
ready - which means the interface is connected and all subscriptions are
already made to mqtt broker
is unsubscribed - when all the subscriptions are removed from the
broker.
This is part of #3 task - adding generated mqtt tests.
The first event (read) was necessery to properly set up test, and start
it when the subscriptions are ready
The second was added for symmetry.

w4bremer commented on templates/apigear/mqtt/mqttbase.h:
```diff
@@ -0,0 +1,47 @@
+    const std::string& getClientId() const;
+
+protected:
+    CWrapper& getCwrapper();
```
I don't like that we hide the shared pointer behind a reference. 
Can't the service, client just directly access the base class cwrapper shared pointer? Why so complicated or do I miss something?

dorotaphanSiili replied:
🤔 I saw it as simplification 😅
 if it is reference then it is kind of promised that it is always there, no need to check for nullptr. I'll change since looks like adds complexity instead of make it simpler

w4bremer commented on templates/module/generated/mqtt/mqttclient.cpp.tpl:
```diff
@@ -1,8 +1,8 @@
 {{- $class := printf "%sClient" .Interface.Name }}
-{{- $pub_interface := printf "I%sPublisher" $interfaceName }}
-{{- $pub_class := printf "%sPublisher" $interfaceName -}}
+{{- $pub_interface := printf "I%sPublisher" ( Camel  $interfaceName ) }}
```
looks like at least one space too much. Would be a good point for clang-format? ;)

w4bremer commented on templates/apigear/mqtt/mqttbase.h:
```diff
@@ -0,0 +1,47 @@
+class APIGEAR_MQTT_EXPORT Base
+{
+public:
+    explicit Base(const std::string& clientID);
```
Maybe, give it a better name as e.g. `CommonBase`  or similar. It basically wraps the common mqtt client functionality. 

Could it be better to do this via composition than inheritance? I don't think we don't to expose all those functions on the Service/Client wrappers

dorotaphanSiili replied:
I'll think of better name

I don't think composition is better option here, the Client and Service would need to repeat whole public interface of the class, those are all functions that the interface adapters will be using

w4bremer commented on templates/apigear/mqtt/baseadapter.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace ApiGear {
+namespace MQTT {
+class APIGEAR_MQTT_EXPORT BaseAdapter
```
I think the file names should match the same pattern as the other classes/files for mqtt.
E.g. `mqttbaseadapter.h`  (or it should not be part of the mqtt namespace)

dorotaphanSiili replied:
I'll change to match current pattern for consistency.
Maybe we should at some point change it? I mean the namespace already says it is mqtt :) if so I'll add a task for later

w4bremer reviewed (changes requested):
It's a bit more than just "tests". The mqtt implementation clearly lacks an initial architecture setup which is on me.
Some of those service, client, base, baseadapters classes probably can be removed with a better architecture.
We probably should plan to do that some time this spring.

Please check my comments, most should be feasible without changing the architecture for now.

dorotaphanSiili committed:
fix: add missing header

dorotaphanSiili committed:
feat(ci): start mosquitto for mqtt tests

dorotaphanSiili committed:
enhance cmake script to have logs from app

w4bremer committed:
fix(mqtt): Resolve minor build issues

Remove explicit links to the MQTT library, since they are provided by
the api module library itself.
Add explicit link to the utilities library.

dorotaphanSiili committed:
fix(mqtt): clean up on disconnected

the on disconnected callbacks were moved to when the mqtt client states
it has disconnected (which is slightly later then disconnect function's
end.
The thread to reset the pointer and reconnect is starting now only to
when it really is needed. It used to be needed for all stituations to
destroy mqtt client pointer with paho functions, but now with the
wrapper is will be cleaned up anyway.
